### PR TITLE
TYP: misc return types

### PR DESCRIPTION
**Pull request by twoertwein**

### CODE REVIEW

File: pandas/core/frame.py [hunk 1]
```diff
@@ -11815,19 +11899,19 @@ def kurt(
     product = prod
 
     @doc(make_doc("cummin", ndim=2))
-    def cummin(self, axis: Axis | None = None, skipna: bool = True, *args, **kwargs):
+    def cummin(self, axis: Axis = 0, skipna: bool = True, *args, **kwargs) -> Self:
```
Comment (twoertwein) [inline pandas/core/frame.py:11902]: `None` is not documented and `0` seems to behave the same?!

File: pandas/core/generic.py [hunk 2]
```diff
@@ -478,8 +478,9 @@ def _validate_dtype(cls, dtype) -> DtypeObj | None:
     # ----------------------------------------------------------------------
     # Construction
 
+    # error: Signature of "_constructor" incompatible with supertype "PandasObject"
     @property
-    def _constructor(self) -> Callable[..., Self]:
+    def _constructor(self) -> Callable[..., Self]:  # type: ignore[override]
```
Comment (twoertwein) [inline pandas/core/generic.py:483]: Should technically be `type[Self]` but the NDFrame constructor allows far fewer arguments than the Series/DataFrame constructor, which leads to many errors when `_constructor()(...)` is called inside NDFrame, which are at runtime only reached by sub-classes.

File: pandas/_typing.py [hunk 3]
```diff
@@ -118,6 +118,7 @@
     Concatenate: Any = None
 
 HashableT = TypeVar("HashableT", bound=Hashable)
+HashableT2 = TypeVar("HashableT2", bound=Hashable)
```
Comment (mroeschke) [inline pandas/_typing.py:121]: Curious why we need a second one?

Comment (twoertwein) [reply pandas/_typing.py:121]: needed for `_from_nested_dict`. When using the same TypeVar in a function signature, the type checkers bind these to the same type:

```py
def (x: HashableT) -> HashableT: ...

x(None) # return None
x((1, 2, 3)) # return tuple[int, int, int]
```

When we have a function that tries to bind two different sets of types, we need a new TypeVar

`def (x: HashableT, y: HashableT2) -> dict[HashableT, dict[HashableT2, Any]]: ...`

(pandas-stubs uses the name HashableT2)

**Review by WillAyd — APPROVED** [review]

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @twoertwein

**Pull request merged**